File: p79.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool search(vector<vector<char>>& board, string& word, int i, int row, int col) {
        // out of bounds
        if(row >= board.size() || row < 0)
            return false;
        if(col >= board[0].size() || col < 0)
            return false;
        // found the word
        if(i == word.size()-1 && board[row][col] == word[word.size()-1])
            return true;
        // not the right letter or already visited
        if(board[row][col] != word[i])
            return false;
        // fill it so we don't cycle back
        char c = board[row][col];
        board[row][col] = ' ';
        bool answer = search(board, word, i+1, row, col-1) || search(board, word, i+1, row, col+1)|| search(board, word, i+1, row-1, col) || search(board, word, i+1, row+1, col); 
        // reset the cell
        board[row][col] = c;
        return answer;
    }
    bool exist(vector<vector<char>>& board, string word) {
        for(int i = 0; i < board.size(); ++i) {
            for(int j = 0; j < board[0].size(); ++j) {
                if(search(board, word, 0, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }
};

File: p79.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        visited = set()
        def dfs(i, j, idx):
            if idx > len(word)-1:
                return True
            if board[i][j] == word[idx]:
                if idx == len(word)-1:
                    return True
                if i-1 >= 0 and (i-1, j) not in visited:
                    visited.add((i-1, j))
                    ret = dfs(i-1, j, idx+1)
                    if ret:
                        return True
                    visited.remove((i-1, j))
                if i+1 < len(board) and (i+1, j) not in visited:
                    visited.add((i+1, j))
                    ret = dfs(i+1, j, idx+1)
                    if ret:
                        return True
                    visited.remove((i+1, j))
                if j-1 >= 0 and (i, j-1) not in visited:
                    visited.add((i, j-1))
                    ret = dfs(i, j-1, idx+1)
                    if ret:
                        return True
                    visited.remove((i, j-1))
                if j+1 < len(board[0]) and (i, j+1) not in visited:
                    visited.add((i, j+1))
                    ret = dfs(i, j+1, idx+1)
                    if ret:
                        return True
                    visited.remove((i, j+1))
            return False

        for i in range(len(board)):
            for j in range(len(board[0])):
                visited.add((i, j))
                ret = dfs(i, j, 0)
                visited.remove((i, j))
                if ret:
                    return True

        return False
